pkg/resource/providerconfig.go

```diff
@@ -46,6 +56,119 @@ func IsMissingReference(err error) bool {
 	return ok
 }
 
+// A CredentialExtractor extracts a provider's credentials.
+type CredentialExtractor interface {
+	Extract(context.Context, string, interface{})
+}
+
+// An ExtractionHandlerFn is a function that can extract a specific type of
+// credentials.
+type ExtractionHandlerFn func(ctx context.Context, c interface{}) ([]byte, error)
```

Comment:
Does this take `interface{}` instead of `xpv1.CredentialsSelectors` because we expect that some `ExtractionHandlerFn` functions would be built using a different kind of struct as input? If so we might consider narrowing it down a bit by using our own interface. e.g.:

```go
type CredentialSelector interface {
    // A dummy method that types can implement to indicate they're a credential selector.
    CredentialSelector() bool
}
```

I think this could help with readability/code documentation.

Reply:
I ended up just passing the `CommonCredentialSelectors` to the builder functions, which feels a little weird, but enables folks to define their custom functions with other struct types as well without having to use the interface method defined above. Would be interested to hear if you have a preference one way or the other?